src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -1488,7 +1488,7 @@ bool AbstractWriteLog<I>::check_allocation(
       no_space = true; /* Entries must be retired */
     }
     /* Don't attempt buffer allocate if we've exceeded the "full" threshold */
-    if (m_bytes_allocated + bytes_allocated > m_bytes_allocated_cap) {
+    if (m_bytes_allocated + bytes_allocated >= m_bytes_allocated_cap) {
```

Comment:
I don't understand this bit.  In the commit message you are saying
> the number of bytes allocated should be less than the total capacity, not equal to, in order to distinguish full cache from empty cache..

but it seems bogus to me.  An empty cache would have `m_bytes_allocated == 0` so I don't see a problem with `m_bytes_allocated + bytes_allocated == m_bytes_allocated_cap`.

Reply:
regard space as a ring. 
when m_bytes_allocated == 0, m_first_free_entry == m_first_valid_entry. 
when  m_bytes_allocated == m_bytes_allocated_cap, m_first_free_entry == m_first_valid_entry.
The two will be treated as the same, unable to distinguish between empty and full. So should keep one for free like this in rwl:
https://github.com/ceph/ceph/blob/d3d967794cf48dd0f14f0e76fcbd3e40f3f41436/src/librbd/cache/pwl/rwl/WriteLog.cc#L315

Reply:
In this commit: https://github.com/ceph/ceph/pull/41354/commits/ea65553b4a9ee1349c6da8452d861afe579e99e9
log entries was deleted in ssd mode. But there's no new way to prevent cache from filling up.
`this->m_free_log_entries = new_root->num_log_entries - 1;`

Reply:
Can you explain the problem with filling up the cache to capacity?  I don't see anything wrong with `m_bytes_allocated == m_bytes_allocated_cap && m_first_free_entry == m_first_valid_entry`.

Reply:
The entry design idea of PWL is: if `m_first_free_entry == m_first_valid_entry`, the cache is empty. see: https://github.com/ceph/ceph/blob/c42712dd18032d9651e2ca5f6fb9ae5a078378df/src/librbd/cache/pwl/AbstractWriteLog.h#L281
And in public code, when load cache, if `m_first_free_entry == m_first_valid_entry`, set  `m_cache_state->empty = true;` see:  https://github.com/ceph/ceph/blob/c42712dd18032d9651e2ca5f6fb9ae5a078378df/src/librbd/cache/pwl/AbstractWriteLog.cc#L544
In rwl part, also use this when load:
https://github.com/ceph/ceph/blob/c42712dd18032d9651e2ca5f6fb9ae5a078378df/src/librbd/cache/pwl/rwl/WriteLog.cc#L355
In ssd part, this logic had been deleted.

Reply:
Yes, it was deleted because it was broken.  But I see your point -- we do need to mimic it in `ssd` mode.

I think a better fix would be to set `ssd` mode `m_bytes_allocated_cap` to `m_log_pool_size - DATA_RING_BUFFER_OFFSET - MIN_WRITE_ALLOC_SSD_SIZE`.  This way `AbstractWriteLog` wouldn't grow an `ssd` specific workraround.

Reply:
> Yes, it was deleted because it was broken. But I see your point -- we do need to mimic the `num_log_entries - 1` logic in `ssd` mode.
> 
> I think a better fix would be to set `ssd` mode `m_bytes_allocated_cap` to `m_log_pool_size - DATA_RING_BUFFER_OFFSET - MIN_WRITE_ALLOC_SSD_SIZE`. This way `AbstractWriteLog` wouldn't need an `ssd` specific workraround.

Agree!